support-services/src/main/scala/com/gu/zuora/ZuoraGiftService.scala

```diff
@@ -77,37 +77,47 @@ class ZuoraGiftService(val config: ZuoraConfig, client: FutureHttpClient)(implic
     } yield successOrFailureResponse).value
 
     val subscriptionNumber = subscription.subscriptionNumber
-    // For the revenue recognition we want to continue even if there is a failure because by this point
-    // the redemption will have succeeded so we need to let the user know about that and then sort out the
-    // revenue recognition separately
-    response.transform{
+    response.transform {
       case Success(Some(response)) if response.success =>
         SafeLogger.info(s"Successfully set up revenue recognition for Digital Subscription gift $subscriptionNumber")
         Success(())
-      case Success(Some(response)) =>
-        val message = response.errorMessage.getOrElse("unknown")
-        val errorMessage = scrub"Failed to set up revenue recognition for Digital Subscription gift $subscriptionNumber. Error was ${message}"
-        SafeLogger.error(errorMessage) //TODO: Add an alarm
-        Success(())
-      case Failure(originalError) =>
-        val errorMessage = scrub"Failed to set up revenue recognition for Digital Subscription gift $subscriptionNumber."
-        SafeLogger.error(errorMessage, originalError) //TODO: Add an alarm
-        Success(())
-      case _ =>
-        val errorMessage = scrub"Failed to set up revenue recognition for Digital Subscription gift $subscriptionNumber."
+      case failed =>
+        val errorMessage = scrub"Failed to set up revenue recognition for Digital Subscription gift $subscriptionNumber. Error was $failed"
         SafeLogger.error(errorMessage) //TODO: Add an alarm
+        // For the revenue recognition we want to continue even if there is a failure because by this point
+        // the redemption will have succeeded so we need to let the user know about that and then sort out the
+        // revenue recognition separately
         Success(())
     }
   }
 
   private def getRevenueSchedule(ratePlanChargeId: String): Future[RevenueSchedulesResponse] =
     get[RevenueSchedulesResponse](s"revenue-schedules/subscription-charges/${ratePlanChargeId}", authHeaders)
 
+  private def maybeDistributeRevenue(
+    revenueSchedule: RevenueSchedulesResponse,
+    termDates: TermDates
+  ) =
+    if (revenueScheduleIsAlreadyDistributed(revenueSchedule))
+      OptionT.fromOption(Some(ZuoraSuccessOrFailureResponse(success = true, None)))
+    else
+      for {
+        revenueScheduleNumber <- OptionT.fromOption(revenueSchedule.revenueSchedules.headOption.map(_.number))
+        successOrFailureResponse <- OptionT.liftF(distributeRevenueByStartAndEndDates(revenueScheduleNumber, termDates))
+      } yield successOrFailureResponse
+
+  private def revenueScheduleIsAlreadyDistributed(revenueSchedule: RevenueSchedulesResponse) =
+    revenueSchedule.revenueSchedules.exists(revenueSchedule =>
+      // revenueSchedule.amount and revenueSchedule.undistributedUnrecognizedRevenue will be
+      // the same until the distribute revenue call has been made
```

Comment:
I presume zuora otherwise would automatically recognise the revenue over the charge dates.  Did you have to change the product charge so that the revenue is not distributed by default?

Reply:
Yes you set a 'Billing manual' revenue recognition rule on the product charge and that tells Zuora to keep all revenue as 'Undistributed' until it is manually distributed, either through the UI or via and api call